samples/99_zil_interpreter/app/builtins.rb

```diff
@@ -37,9 +37,25 @@ def expect_minimum_argument_count!(arguments, min_count)
 ZIL_BUILTINS[:LVAL] = define_for_evaled_arguments { |arguments, context|
   expect_argument_count!(arguments, 1)
   var_atom = arguments[0]
-  raise FunctionError, "No local value for #{var_atom.inspect}" unless context.locals.key? var_atom
 
-  context.locals[var_atom]
+  result = :ZIL_LOCAL_VALUE_NOT_ASSIGNED
+  [context.locals, *context.locals_stack].each { |stack|
+    if stack.key? var_atom
+      result = stack[var_atom]
+      break
+    end
+  }
+
+  raise FunctionError, "No local value for #{var_atom.inspect}" if result == :ZIL_LOCAL_VALUE_NOT_ASSIGNED
+  result
+}
+
+ZIL_BUILTINS[:GVAL] = define_for_evaled_arguments { |arguments, context|
+  expect_argument_count!(arguments, 1)
+  var_atom = arguments[0]
+  raise FunctionError, "No global value for #{var_atom.inspect}" unless context.globals.key? var_atom
+
+  context.globals[var_atom]
 }
 
 ZIL_BUILTINS[:VALUE] = define_for_evaled_arguments { |arguments, context|
```

Comment:
I think VALUE must also be adapted to use the stack

Reply:
Updated :VALUE.  Added unit test `test_builtin_value`.